company-eudc.el

```diff
@@ -0,0 +1,155 @@
+EUDC is the Emacs Unified Directory Client, a common interface to
+directory servers and contact information.
+
+Completion will be attempted in the To, Cc, Bcc, From, and
+Reply-To header fields of message buffers only. I.e. you can not
+use this backend for completing email addresses outside a message
+header.
+
+To load the `company-eudc' backend, simply do
+
+    (require 'company-eudc)
+
+With this alone, no completions from EUDC will be offered
+yet. You will additionally need to decide how you want to use
+it. Since EUDC can query remote resources (such as e.g. LDAP
+servers), the completion process may take a while. Depending on
+how you have configured `company-mode', this may result in Emacs
+being blocked for extensive periods of time. To accommodate this,
+the `company-eudc' backend can either install itself into the
+list of `company-mode' backends for `message-mode' (see
+`company-eudc-activate-autocomplete'), or provide autocompletion
+through `company-mode' only when bound to a key (see
+`company-eudc-expand-inline')."
+  (interactive (list 'interactive))
+  (pcase command
+    (`interactive (company-begin-backend 'company-eudc))
+    (`prefix (and (memq major-mode company-eudc-modes)
+		  (let ((case-fold-search t))
+		    (looking-back
+		     "^\\([^ :]*-\\)?\\(To\\|B?Cc\\|From\\|Reply-to\\):.*? *\\([^,;]*\\)"
+		     (line-beginning-position)))
+		  (company-grab-symbol)))
+    (`candidates (let* ((q-result (eudc-query `((name . ,arg)))))
+		   (cl-loop for person-record in q-result
+			    collect (company-eudc-email-address-from-alist
+				     person-record))
+		  ))
+    (`sorted t)))
+
+;;;###autoload
+(defun company-eudc-activate-autocomplete ()
+  "Provide `company-eudc' completions under `company-mode' control.
+This function installs `company-eudc' in the list of
+`company-mode' backends for `message-mode'. Completion candidates
+from EUDC will thus be offered by `company-mode' as any other
+candidate.
+
+To get this behaviour, do
+
+    (require 'company-eudc)
+    (company-eudc-activate-autocomplete)
+
+If you have configured many and/or slow servers for EUDC, this
+will block Emacs for some time (i.e. until EUDC has delivered its
+results). If this is a frequent issue, and you would like to
+avoid this, use `company-eudc-expand-inline' instead. "
+  (interactive)
+  (add-to-list 'company-backends 'company-eudc)
+  (with-eval-after-load "message"
+    (add-hook 'message-mode-hook
+	      (lambda () (add-to-list 'company-backends 'company-eudc)))))
+
+;;;###autoload
+(defun company-eudc-expand-inline ()
+  "Provide `company-eudc' completions in `message-mode' interactively.
+This function triggers `company-mode' completion at point, using
+the `company-eudc' backend only. It is intended for being bound
+to a key chord; for example:
+
+    (require 'company-eudc)
+    (with-eval-after-load \"message\"
+      (define-key message-mode-map (kbd \"<C-tab>\") 'company-eudc-expand-inline))
```

Comment:
Why not bind `company-eudc` to a key directly?

This function seems unnecessary (though the explanation in the docstring is most likely useful).